mgmtsystem/scvmm.py

```diff
@@ -174,31 +230,52 @@ def wait_vm_suspended(self, vm_name, num_sec=300):
+
             num_sec=num_sec)
 
-    def clone_vm(self, source_name, vm_name):
-        """It wants exact host and placement (c:/asdf/ghjk) :("""
-        raise NotImplementedError('clone_vm not implemented.')
+    def clone_vm(self, vm_source, vm_host, path, vm_name):
+        script = """
+        $vm_new = Get-SCVirtualMachine -Name "{vm_source}" -VMMServer $scvmm_server
+        $vm_host = Get-SCVMHost -VMMServer $scvmm_server -ComputerName "{vm_host}"
+        New-SCVirtualMachine -Name "{vm_name}" -VM $vm_new -VMHost $vm_host -Path "{path}" -StartVM
+        """.format(vm_name=vm_name, vm_source=vm_source, vm_host=vm_host, path=path)
+        self.logger.info(" Deploying SCVMM VM `{}` from Clone of `{}`"
+            .format(vm_name, vm_source))
+        self.run_script(script)
 
     def does_vm_exist(self, vm_name):
         result = self.run_script("Get-SCVirtualMachine -Name \"{}\" -VMMServer $scvmm_server"
             .format(vm_name)).strip()
         return len(result) > 0
 
-    def deploy_template(self, template, vm_name=None, host_group=None, **bogus):
+    def does_template_exist(self, template):
+        result = self.run_script("Get-SCVMTemplate -Name \"{}\" -VMMServer $scvmm_server"
+            .format(template)).strip()
+        return len(result) > 0
+
+    def deploy_template(self, template, vm_host, host_group, vm_name=None):
         script = """
         $tpl = Get-SCVMTemplate -Name "{template}" -VMMServer $scvmm_server
-        $vmhostgroup = Get-SCVMHostGroup -Name "{host_group}" -VMMServer $scvmm_server
-        $vmc = New-SCVMConfiguration -VMTemplate $tpl -Name "{vm_name}" -VMHostGroup $vmhostgroup
+        $vm_host_group = Get-SCVMHostGroup -Name "{host_group}" -VMMServer $scvmm_server
+        $vmc = New-SCVMConfiguration -VMTemplate $tpl -Name "{vm_name}" -VMHostGroup $vm_host_group
         Update-SCVMConfiguration -VMConfiguration $vmc
-        New-SCVirtualMachine -Name "{vm_name}" -VMConfiguration $vmc #-VMMServer $scvmm_server
-        """.format(template=template, vm_name=vm_name, host_group=host_group)
+        New-SCVirtualMachine -Name "{vm_name}" -VMConfiguration $vmc
+        """.format(template=template, vm_name=vm_name, vm_host=vm_host, host_group=host_group)
         self.logger.info(" Deploying SCVMM VM `{}` from template `{}` on host group `{}`"
             .format(vm_name, template, host_group))
         self.run_script(script)
         self.start_vm(vm_name)
         return vm_name
 
+    def mark_as_template(self, vm_name, library, library_share):
+        script = """
+        $VM = Get-SCVirtualMachine -Name \"{vm_name}\" -VMMServer $scvmm_server
+        New-SCVMTemplate -Name \"{vm_name}-tpl\" -VM $VM -LibraryServer \"{ls}\" -SharePath \"{lp}\"
+         """.format(vm_name=vm_name, ls=library, lp=library_share)
+        self.logger.info(" Creating SCVMM Template `{vm_name}` from VM `{vm_name}`-tpl"
+            .format(vm_name=vm_name))
+        self.run_script(script)
```

Comment:
So, here what it should do is that it should rename the original VM (testvm) to eg (testvm_copytemplate), then it should create a template (testvm) from (testvm_copytemplate) and delete the (testvm_copytemplate) so in the end, only the template exists and has the same name as the original VM. Basically just two more calls wrapping this script call :)